File: isegoria/src/main/java/com/eulersbridge/isegoria/feed/photos/PhotoViewHolder.java
package com.eulersbridge.isegoria.feed.photos;

import android.content.Context;
import android.support.annotation.Nullable;
import android.view.View;
import android.widget.ImageView;

import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
import com.eulersbridge.isegoria.GlideApp;
import com.eulersbridge.isegoria.R;
import com.eulersbridge.isegoria.network.api.models.Photo;
import com.eulersbridge.isegoria.util.ui.LoadingAdapter;

class PhotoViewHolder extends LoadingAdapter.ItemViewHolder<Photo> {

    interface ClickListener {
        void onClick(Context context, int position);
    }

    private final ClickListener clickListener;
    final private ImageView imageView;

    private boolean isImageLoadStarted = false;

    PhotoViewHolder(View itemView, ClickListener clickListener) {
        super(itemView);

        this.clickListener = clickListener;

        imageView = itemView.findViewById(R.id.photo_grid_item_image_view);
        imageView.setOnClickListener(view -> {
            if (this.clickListener != null)
                this.clickListener.onClick(imageView.getContext(), getAdapterPosition());
        });
    }

    @Override
    protected void onRecycled() {
        if (imageView.getContext() != null && isImageLoadStarted)
            GlideApp.with(imageView.getContext()).clear(imageView);
    }

    @Override
    protected void setItem(@Nullable Photo item) {
        if (item != null) {
            imageView.setContentDescription(item.title);

            GlideApp.with(imageView.getContext())
                    .load(item.thumbnailUrl)
                    .transition(DrawableTransitionOptions.withCrossFade())
                    .into(imageView);

            isImageLoadStarted = true;
        }
    }
}

File: isegoria/src/main/java/com/eulersbridge/isegoria/feed/photos/PhotoViewHolder.kt
package com.eulersbridge.isegoria.feed.photos

import android.content.Context
import android.view.View
import android.widget.ImageView

import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions
import com.eulersbridge.isegoria.GlideApp
import com.eulersbridge.isegoria.R
import com.eulersbridge.isegoria.network.api.models.Photo
import com.eulersbridge.isegoria.util.ui.LoadingAdapter

internal class PhotoViewHolder(itemView: View, private val clickListener: ClickListener?) : LoadingAdapter.ItemViewHolder<Photo>(itemView) {

    private val imageView: ImageView = itemView.findViewById(R.id.photo_grid_item_image_view)
    private var isImageLoadStarted = false

    internal interface ClickListener {
        fun onClick(context: Context, position: Int)
    }

    init {
        imageView.setOnClickListener {
            clickListener?.onClick(imageView.context, adapterPosition)
        }
    }

    override fun onRecycled() {
        if (imageView.context != null && isImageLoadStarted)
            GlideApp.with(imageView.context).clear(imageView)
    }

    override fun setItem(item: Photo?) {
        if (item != null) {
            imageView.contentDescription = item.title

            GlideApp.with(imageView.context)
                    .load(item.thumbnailUrl)
                    .transition(DrawableTransitionOptions.withCrossFade())
                    .into(imageView)

            isImageLoadStarted = true
        }
    }
}